be/src/storage/binlog_reader.cpp

```diff
@@ -0,0 +1,175 @@
+        }
+        return status;
+    }
+
+    // End of segment file, and switch to next log entry
+    status = _binlog_file_reader->next();
+    if (!status.ok() && !status.is_end_of_file()) {
+        return status;
+    }
+
+    if (status.is_end_of_file()) {
+        RETURN_IF_ERROR(_seek_to_file_meta(_next_version, _next_seq_id));
+    }
+    RETURN_IF_ERROR(_seek_to_segment_row(_next_seq_id));
+
+    return get_next(chunk, max_version_exclusive);
+}
+
+Status BinlogReader::_seek_to_file_meta(int64_t version, int64_t seq_id) {
+    auto status_or = _binlog_manager->find_binlog_file(version, seq_id);
+    if (!status_or.ok()) {
+        return status_or.status();
+    }
+    _file_meta = status_or.value();
+    std::string file_name = _binlog_manager->binlog_file_name(_file_meta->id());
+    _binlog_file_reader = std::make_shared<BinlogFileReader>(file_name, _file_meta);
+    RETURN_IF_ERROR(_binlog_file_reader->seek(version, seq_id));
+    return Status::OK();
+}
+
+Status BinlogReader::_seek_to_segment_row(int64_t seq_id) {
+    _log_entry_info = _binlog_file_reader->log_entry();
+    LogEntryTypePB log_entry_type = _log_entry_info->log_entry->entry_type();
+    if (log_entry_type == EMPTY) {
+        return Status::OK();
+    }
+    CHECK_EQ(log_entry_type, INSERT_RANGE) << "currently only support INSERT_RANGE";
+    _next_seq_id = seq_id;
+    CHECK(_log_entry_info->start_seq_id <= _next_seq_id)
+            << "Seek to invalid seq, start_seq_id " << _log_entry_info->start_seq_id
+            << ", target seq_id " << _next_seq_id;
+    int32_t start_row_id = seq_id - _log_entry_info->start_seq_id + _log_entry_info->start_row_id;
+    return _init_segment_iterator(start_row_id);
+}
+
+Status BinlogReader::_init_segment_iterator(int32_t start_row_id) {
+    if (_segment_iterator != nullptr) {
+        _segment_iterator->close();
+        _segment_iterator.reset();
+    }
+
+    RowsetIdPB* rowset_id_pb = _log_entry_info->file_id->mutable_rowset_id();
+    RowsetId rowset_id;
+    rowset_id.init(rowset_id_pb->hi(), rowset_id_pb->mi(), rowset_id_pb->lo());
+    if (_rowset != nullptr && rowset_id != _rowset->rowset_id()) {
+        _rowset->release();
+        _rowset.reset();
+    }
+    if (_rowset == nullptr) {
+        _rowset = _binlog_manager->get_rowset(rowset_id);
+        _rowset->acquire();
+        RETURN_IF_ERROR(_rowset->load());
+    }
+
+    int seg_index = _log_entry_info->file_id->segment_index();
+    SegmentSharedPtr seg_ptr = _rowset->segments()[seg_index];
+    vectorized::SegmentReadOptions seg_options;
+    ASSIGN_OR_RETURN(seg_options.fs, FileSystem::CreateSharedFromString(_rowset->rowset_path()));
+    seg_options.chunk_size = _chunk_size;
+    vectorized::SparseRange range(start_row_id, seg_ptr->num_rows());
+    seg_options.rowid_range_option =
+            std::make_shared<vectorized::RowidRangeOption>(_rowset->rowset_id(), seg_index, range);
+    auto res = seg_ptr->new_iterator(_schema, seg_options);
```

Comment:
Can the same `_segment_iterator ` (the same segment file) be initialized multi time? maybe reuse _segment_iterator later?

Reply:
Does `new_iterator `  open other index files other than segment file?